README.md
# Rotkehlchen

<img src="https://raw.githubusercontent.com/rotkehlchenio/rotkehlchen/master/ui/images/rotkehlchen_no_text.png" width="250">

Rotkehlchen is an asset management and accounting application specializing in Crypto assets and aims to also help with tax reporting. It has integration for multiple exchanges with more coming soon.

## Table of Contents

- [Disclaimer](#disclaimer)
- [Installation](#installation)
    - [Packaged binaries](#packaged-binaries)
        - [Linux](#linux)
        - [OSX](#osx)
    - [Build from Source](#build-from-source)
- [Usage](#usage)
    - [Signin/Signup](#signinsignup)
    - [Changing the Profit Currency](#changing-the-profit-currency)
    - [Connecting to an Ethereum Client](#connecting-to-an-ethereum-client)
    - [Adding Exchanges](#adding-exchanges)
    - [Adding Fiat Balances](#adding-fiat-balances)
    - [Adding And Removing Blockchain Accounts](#adding-and-removing-blockchain-accounts)
    - [Adding And Removing Ethereum Tokens](#adding-and-removing-ethereum-tokens)
    - [Manually Adding Trades Or Taxable Events](#manually-adding-trades-or-taxable-events)
    - [Creating a Tax Report](#creating-a-tax-report)
- [Wanna Help?](#wanna-help)
    - [Contribute](#contribute)
    - [Get Premium](#get-premium----coming-soon)
    - [Donate](#donate)

## Disclaimer

ROTKEHLCHEN'S SERVICE IS INTENDED ONLY TO IMPORT, DISPLAY, STORE, AND GENERATE REPORTS AND STATISTICS FROM CRYPTOCURRENCY TRANSACTION DATA AND OTHER ASSETS. YOU, AS THE USER OF THE SERVICE, ARE RESPONSIBLE FOR THE TRANSACTION DATA IMPORTED, STORED, AND ANALYZED BY THE SERVICE. IF INACCURATE OR INCOMPLETE TRANSACTION DATA IS PROVIDED TO THE SERVICE, THE REPORTS GENERATED BY THE SERVICE WILL ALSO BE INACCURATE OR INCOMPLETE.

ROTKEHLCHEN DOES NOT PROVIDE TAX OR LEGAL ADVICE THROUGH THE SERVICE AND THE SERVICE IS NOT INTENDED TO BE A SUBSTITUTE FOR PROFESSIONAL TAX OR LEGAL ADVICE OR FOR YOUR OWN DILIGENCE IN ENSURING THE ACCURACY OF ANY AND ALL INFORMATION YOU SUBMIT TO TAX AUTHORITIES. IT IS YOUR RESPONSIBILITY TO PRESENT YOUR TRANSACTION DATA AND ANY REPORTS GENERATED BY THE SERVICE TO YOUR TAX OR LEGAL ADVISORS FOR REVIEW. ALWAYS SEEK THE ADVICE OF YOUR ACCOUNTANTS, ATTORNEYS, OR OTHER QUALIFIED TAX ADVISORS WITH ANY QUESTIONS YOU MAY HAVE REGARDING YOUR TAX LIABILITY. NEVER DISREGARD PROFESSIONAL TAX OR LEGAL ADVICE OR DELAY IN SEEKING IT BECAUSE OF SOMETHING YOU HAVE READ OR BEEN PROVIDED FROM THE SERVICE.

THE SERVICE IS NOT INTENDED FOR CONTINUOUS MONITORING OF TRANSACTION DATA OR FOR AUTOMATIC TRANSMISSION OF YOUR TRANSACTION DATA OR ANY REPORTS OR SAMPLE FORMS GENERATED BY THE SERVICE TO ANY TAX AUTHORITY. YOU CANNOT SUBMIT PAYMENT TO ANY TAX AUTHORITY THROUGH THE SERVICE, AND WE ARE NOT RESPONSIBLE FOR ANY PAYMENT OR NON-PAYMENT OF TAXES BY YOU.

## Installation

The easiest way to start rotkehlchen is to download the packaged binary for your Operating system. For now only Linux and OSX is supported. To see how to do this go to the [next section](#packaged-binaries).

### Packed Binaries

#### Linux

Go to the [releases](https://github.com/rotkehlchenio/rotkehlchen/releases) page and download the linux-x64 package from the [latest release](https://github.com/rotkehlchenio/rotkehlchen/releases/latest).

Unzip it in a directory of your choice. In the root directory of the unzipped archive there is a `rotkehlchen` executable. Run it to start rotkehlchen.

#### OSX

Go to the [releases](https://github.com/rotkehlchenio/rotkehlchen/releases) page and download the darwin-x64 package from the [latest release](https://github.com/rotkehlchenio/rotkehlchen/releases/latest).

Unzip it in a directory of your choice. In the root directory of the unzipped archive there is a `rotkehlchen` executable. Run it to start rotkehlchen.

### Build from Source

You can also build rotkehlchen in OSX and Linux from source. In order to see how to do that go to [Build from Source](docs/source_install.md).

## Usage

In this section we are going to see how to use various parts of the Rotkehlchen application.

### Sign-in/Signup

When you start Rotkehlchen you are greeted with a sign-in/signup prompt.

If you have never created an account before press "Create New Account". Provide a username and a password. Don't forget this password. It will be used to encrypt all your local files. If you have purchased a premium subscription you can also add the API Key and secret here.

If you already have an account just write the name and the password here.

All accounts are created in the rotkehlchen directory which is located in: `$HOME/.rotkehlchen`. Each user has their own subdirectory.

### Changing the Profit Currency

Rotkehlchen calculates everything, including your profit/loss for your tax report into a given FIAT currency. This is what we call the `profit_currency`. By default this is the US Dollar. You can easily change this setting by clicking on the currency icon the top right menu and changing it to the currency you are using.

### Connecting to an Ethereum Client

When Rotkehlchen begins it tries to connect to a local ethereum node running with an rpc port set at the default port `8545`. If no client is running then all blockchain queries will use etherscan and this will be rather slower.

If you want to connect to an ethereum client running in a non-default port you can set that through the settings. Click the person icon on the top right menu and select "Settings" from the drop down menu. This will take you to the settings page. Write the desired port in the ETH RPC Port textbox

### Adding Exchanges

You can integrate many different exchanges with Rotkehlchen. Currently supported exchanges are: Kraken, Poloniex, Bittrex, Bitmex and Binance.

To do so you have to go to your exchange and create an API key. If the exchange allows it make sure that the API Key only has reading/querying permissions to your account and nothing else since that is all the permissions needed by Rotkehlchen.

Click the Person Icon on the top right menu and then choose "User Settings". This will take you to your user settings page. Under the exchange settings sections, select your exchange from the dropdown menu. Then copy and paste the `API Key` and the `API Secret` in the respective text fields and press submit.

If all went well, then you will get a confirmation that the connection was successful. If not please doublecheck that the key and secret are correct.

### Adding Fiat Balances

Rotkehlchen is an asset management application. Thus you can track all your assets in one place including the FIAT balances you have.

To add or modify the amount of an owned FIAT currency, go to the user settings page, scroll down to the "Fiat Balances" section and choose the currency from the dropdown menu. Input the modified balance in the text box and press the Modify button.

### Adding and Removing Blockchain Accounts

Rotkehlchen allows to track balances of blockchain accounts.

To track an account go to the user settings page, scroll down to the "Blockchain Balances" section and choose the blockchain from the dropdown menu. For now only Bitcoin and Ethereum chains are supported. Then type or paste the address in the "Account" textbox and press the "Add" Button.

To stop tracking a particular account scroll down to the accounts tables and simply right click on the account you want to stop tracking and select "Delete" from the context menu.

### Adding and Removing Ethereum Tokens

Rotkehlchen will check all of your ethereum accounts for balances of a given list of tokens. This list can be provided and modified by you.

To do so go to the user settings page, scroll down to the "Blockchain Balances" section and look at the "Select Eth Tokens to Track" multiselection widget.

In order to add a token to the tracked tokens find it on the left menu and click it so that it gets added to your tokens.

In order to stop tracking a tocken, find it on the right menu and click it to stop tracking it.

All account balances will be updated at this point.

### Manually Adding Trades Or Taxable Events

Rotkehlchen will pull all your trade history from the exchanges whenever it needs it. But most of us have probably also done some OTC trades or taxable events at some point. Such events could even just be mining tokens, depending on your jurisdiction, participating in an ICO or getting paid in crypto.

On the left menu click on the trades button and select "OTC Trades" from the dropdown menu. This will take you to the OTC Trades page.

To add a new trade or taxable event, fill in all the field and press the "Add Trade" button.

Some very important things to note. All pairs should be in the form of `BASECURRENCY_QUOTECURRENCY`. For a `buy` this means you are buying `amount` of the `BASE` currency at a price of `rate` `QUOTE` currency per `BASE`. For a `sell` this means you are selling `amount` of the `BASE` currency at a price of `rate` `QUOTE` currency per `BASE`.

If there was a fee for the trade you should input it in the corresponding box and also enter the currency the fee was paid in. Fee can also be 0.

You can provide additional notes or even links to blockchain explorers for each trade.

At the bottom of this page you can see a table of all your OTC trades.

### Creating a Tax Report

Rotkehlchen creates a tax report for you based on your trades and some settings which you can configure at the user settings page. This is essentially a calculation of profit or loss for all your trades based on the given dates.

Some settings are at the moment hard coded, for the German tax jurisdiction. For example all profit/loss calculation is done for trades on a first-in/first-out basis and profits from selling crypto assets after 1 year are non taxable. These settings will be made adjustable soon.


To create a tax report click on the "Tax Report" button from the left menu. Choose a start and an end date for the report and then click the "Generate Report" button.

The calculation may take some time. Once done you have an overview of the profit/loss for the given period, how much of that is taxable, and how much each taxable event category contributes to the total.

Additionally below the overview you get a table containing all of the taxable events that were taken into account in the calculation along with how much of the `profit_currency` you lost or gained through that event.

Finally you can get a nice CSV export by pressing the "Export CSV" button. This export is meant to be imported into google sheets. Press the button and then choose a directory to write the CSV files to. Once done you can import the CSV files into google sheets via its import menu.

## Wanna Help?


### Contribute

It's an open-source project so help is really appreciated.

Open issues or pull requests to help with development, be active in Github and above all be nice.

In order for your Pull Request to be considered it will need to pass the automated CI tests and you will also need to sign the CLA (Contributor's license agreement).

Please follow these rules for commits:
1. Commits should be just to the point, not too long and not too short.
2. Commits should do one thing, if two commits both do the same thing, that's a good sign they should be combined.
3. Commits should follow a simple format. A max 80 columns title, optionally followed by an empty line separator and then the subject.
4. More info on commit messages [here](https://who-t.blogspot.com/2009/12/on-commit-messages.html)
5. **Never** merge master on the branch, always rebase on master. To delete/amend/edit/combine commits follow [this tutorial](https://robots.thoughtbot.com/git-interactive-rebase-squash-amend-rewriting-history)


### Get Premium -- Coming Soon

Buy a premium subscription to the software and help us with the development while at the same time enjoying all the cool neat features coming to premium subscribers.

### Donate

Don't want or can't afform premium? Still wanna help? We also accept donations.

- Donate ETH or ERC20 tokens to this address: [0x9531c059098e3d194ff87febb587ab07b30b1306](https://etherscan.io/address/0x9531c059098e3d194ff87febb587ab07b30b1306)

- Donate BTC to: 1PfvkW8MC7Ns2y8zn6CE2P2t5f19KF8XiW



CHANGELOG.rst
=========
Changelog
=========

* :release:`0.4.0 <2018-09-23>`
* :feature:`144` Rotkehlchen now starts fully supporting Bitmex and allows querying Bitmex history for tax calculations.
* :bug:`163` Properly handle errors in the tax report calculation and in other asynchronous tasks.
* :bug:`155` Check if the local ethereum node is synced before querying balances from it.
* :feature:`153` Add a ``version`` command to display the rotkehlchen version.
* :bug:`159` Gracefully exit if an invalid argument is provided.
* :bug:`151` If an asset stored at Bittrex does not have a BTC market rotkehlchen no longer crashes.
* :feature:`148` Add icons for all tokens to the UI.
* :feature:`74` Add experimental support for Bitmex. Supporting only simple balance query for now.
* :bug:`135` Fix bug in converting binance sell trades to the common rotkehlchen format
* :bug:`140` Don't log an error if the manual margin file is not found

* :release:`0.3.2 <2018-08-25>`
* :feature:`95` Add a UI widget to display the last time the balance data was saved in the DB.
* :bug:`126` Refuse to generate a new tax report if one is in progress and also clean previous report before generating a new one.
* :bug:`123` Return USD as default main currency if DB is new
* :bug:`101` Catch the web3 exception if using a local client with an out of sync chain and report a proper error in the UI
* :bug:`86` Fixed race condition at startup that could result in the banks balance displaying as NaN.
* :bug:`103` After removing an exchange's API key the new api key/secret input form is now properly re-enabled 
* :bug:`99` Show proper error if kraken or binance api key validation fails due to an invalid key having been provided.

* :release:`0.3.1 <2018-06-25>`
* :bug:`96` Periodic balance data storage should now also work from the UI.

* :release:`0.3.0 <2018-06-24>`
* :feature:`90` Add configuration option for it and periodically save balances data in the database
* :bug:`91` Provide more accurate name for the setting for the date from which historical data starts
* :bug:`89` Many typing bugs were found and fixed
* :bug:`83` Fix a bug that did not allow adding or removing ethereum tokens from the tracker
* :feature:`79` Do not crash with exception if an exchange is unresponsive, but instead warn the user.
* :bug:`77` Fix bug caused by reading `taxfree_after_period` from the database

* :release:`0.2.2 <2018-06-05>`
* :bug:`73` Fixer.io api switched to be subscription based and its endpoints are now locked, so we switch to a different currency converter api.
* :bug:`68` All kraken pairs should now work properly. Users who hold XRP, ZEC, USD, GP, CAD, JPY, DASH, EOSD and USDT in kraken will no longer have any problems.

* :release:`0.2.1 <2018-05-26>`
* :bug:`66` Persist all eth accounts in the database as checksummed. Upgrade all existing DB accounts.
* :bug:`63` Unlocking a user account for an application is no longer slow if you have lots of historical price cache files.
* :bug:`61` Overcome etherscan's limit of 20 accounts per query by splitting the accounts list

* :release:`0.2.0 <2018-05-13>`
* :feature:`51` Add customization for the period of time after which trades are tax free.
* :bug:`50` rotkehlchen --help now works again
* :feature:`45` Add option to customize including crypto to crypto trades.
* :feature:`42` Move the accounting settings to their own page.

* :release:`0.1.1 <2018-04-27>`
* :bug:`37` Fix a bug where adding an ethereum account was throwing an exception in the UI.

* :release:`0.1.0 <2018-04-23>`


renderer.js
const zerorpc = require("zerorpc");
// max timeout is now 25 seconds
let client = new zerorpc.Client({timeout:25, heartbeatInterval: 10000});

function Currency(name, icon, ticker_symbol) {
    this.name = name;
    this.icon = icon;
    this.ticker_symbol = ticker_symbol;
}

let EXCHANGES = ['kraken', 'poloniex', 'bittrex'];

let CURRENCIES = [
    new Currency("United States Dollar", "fa-usd", "USD"),
    new Currency("Euro", "fa-eur", "EUR"),
    new Currency("British Pound", "fa-gbp", "GBP"),
    new Currency("Japanese Yen", "fa-jpy", "JPY"),
    new Currency("Chinese Yuan", "fa-jpy", "CNY"),
];
let default_currency = CURRENCIES[0];
let main_currency = default_currency;
let jobs = ['false', 'false'];
client.connect("tcp://127.0.0.1:4242");



let body = $("body");
body.addClass("loading");

function create_box(id, icon, number, currency_icon) {
    // only show 2 decimal digits
    number = number.toFixed(2);
    var str = '<div class="panel panel-primary"><div class="panel-heading" id="'+id+'"><div class="row"><div class="col-xs-3"><i title="' + id + '" class="fa '+ icon +'  fa-5x"></i></div><div class="col-xs-9 text-right"><div class="huge">'+ number +'</div><div id="status_box_text"><i class="fa '+ currency_icon + ' fa-fw"></i></div></div></div></div><a href="#"><div class="panel-footer"><span class="pull-left">View Details</span><span class="pull-right"><i class="fa fa-arrow-circle-right"></i></span><div class="clearfix"></div></div></a></div>';
    return $(str);
}

function create_exchange_box(exchange, number, currency_icon) {
    css_class = 'exchange-icon-inverted';
    if (exchange == 'poloniex') {
	css_class = 'exchange-icon';
    }
    // only show 2 decimal digits
    number = number.toFixed(2);
    var str = '<div class="panel panel-primary"><div class="panel-heading" id="'+exchange+'_box"><div class="row"><div class="col-xs-3"><i><img title="' + exchange + '" class="' + css_class + '" src="ui/images/'+ exchange +'.png"  /></i></div><div class="col-xs-9 text-right"><div class="huge">'+ number +'</div><div id="status_box_text"><i class="fa '+ currency_icon + ' fa-fw"></i></div></div></div></div><a href="#"><div class="panel-footer"><span class="pull-left">View Details</span><span class="pull-right"><i class="fa fa-arrow-circle-right"></i></span><div class="clearfix"></div></div></a></div>';
    return $(str);
}

function showAlert(type, text) {
    var str = '<div class="alert '+ type +' alert-dismissable"><button type="button" class="close" data-dismiss="alert" aria-hidden="true">&times;</button>'+ text +'</div>';
    $(str).prependTo($("#wrapper"));
}

function create_currency_dropdown(fa_default_icon) {
    var str = '<li class="dropdown"><a class="dropdown-toggle" data-toggle="dropdown" href="#"><i id="current-main-currency" class="fa '+ fa_default_icon + ' fa-fw"></i> <i class="fa fa-caret-down"></i></a><ul class="dropdown-menu dropdown-alerts currency-dropdown"></ul></li>';
    $(str).appendTo($(".navbar-right"));
}

function set_ui_main_currency(currency) {
    $('#current-main-currency').removeClass().addClass('fa ' + currency.icon + ' fa-fw');
}

function add_currency_dropdown(currency) {
    var str = '<li><a id="change-to-'+ currency.ticker_symbol.toLowerCase() +'" href="#"><div><i class="fa '+ currency.icon +' fa-fw"></i> Set '+ currency.name +' as the main currency</div></a></li><li class="divider"></li>';
    $(str).appendTo($(".currency-dropdown"));

    $('#change-to-'+ currency.ticker_symbol.toLowerCase()).bind('click', function() {
	client.invoke("set_main_currency", currency.ticker_symbol, (error, res) => {
	    if(error) {
		showAlert('alert-danger', error);
	    } else {
		set_ui_main_currency(currency);
	    }
	});
    });
}

function finish_job() {
    var finished_one = false;
    for (var i = 0; i < jobs.length; i ++) {
	if (jobs[i] == false) {
	    if (!finished_one) {
		jobs[i] = true;
		finished_one = true;
		continue;
	    } else {
		return;
	    }
	}
    }

    // if we get here it means we finished all jobs
    $('#top-loading-icon').removeClass().addClass('fa fa-check-circle fa-fw');
}

client.invoke("get_initial_settings", (error, res) => {
    if (error || res == null) {
	var loading_wrapper = document.querySelector('.loadingwrapper');
	var loading_wrapper_text = document.querySelector('.loadingwrapper_text');
	console.log("Response was: " + res);
	console.error(error);
	loading_wrapper.style.background = "rgba( 255, 255, 255, .8 ) 50% 50% no-repeat";
	loading_wrapper_text.textContent = "ERROR: Failed to connect to the backend. Check Log";
    } else {
	// set main currency
	console.log("server is ready");
	main_currency = res['main_currency'];
	for (var i = 0; i < CURRENCIES.length; i ++) {
	    if (main_currency == CURRENCIES[i].ticker_symbol) {
		set_ui_main_currency(CURRENCIES[i]);
		main_currency = CURRENCIES[i];
		break;
	    }
	}
	// make separate queries for all registered exchanges
	let exchanges = res['exchanges'];
	for (var i = 0; i < exchanges.length; i++) {
	    jobs.push(false);
	}
	for (var i = 0; i < exchanges.length; i++) {
	    client.invoke("query_exchange_total", exchanges[i], true, function (error, res) {
		if (error || res == null) {
		    console.log("Error at first query of an exchange's balance: " + error);
		} else {
		    create_exchange_box(res['name'], res['total'], main_currency.icon).appendTo($('#leftest-column'));
		    finish_job();
		}
	    });
	}
	body.removeClass("loading");
    }
});

client.invoke("query_blockchain_total", (error, res) => {
    if (error || res == null) {
	console.log("Error at querying blockchain total: " + error);
    } else {
	create_box('blockchain balance', 'fa-hdd-o', res['total'], main_currency.icon).appendTo($('#leftest-column'));
	finish_job();
    }
});
client.invoke("query_banks_total", (error, res) => {
    if (error || res == null) {
	console.log("Error at querying bank total: " + error);
    } else {
	create_box('banks balance', 'fa-university', res['total'], main_currency.icon).appendTo($('#leftest-column'));
	finish_job();
    }
});


create_currency_dropdown(default_currency.icon);
for (var i = 0; i < CURRENCIES.length; i++) {
    add_currency_dropdown(CURRENCIES[i]);
}



// let result = document.querySelector('#result');
// formula.addEventListener('input', () => {
//   client.invoke("calc", formula.value, (error, res) => {
//     if(error) {
// 	console.error(error);
//     } else {
// 	result.textContent = res;
//     }
//   });
// });

// formula.dispatchEvent(new Event('input'));
